### PR TITLE
Tabular train session inserts entry into execution-table

### PR DESCRIPTION
**Pull request by farisdurrani**
# Major Changes
1. Execution-table update: every time the user clicks on Train in tabular, a new entry is made into the execution-table in DynamoDB as a precursor to a realtime user dashboard. For now, there will only be one entry for each train session with a constant status `STARTED` no matter the execution success. Working for #371 

image
The execution_id is made using the timestamp and the user ID.

# Minor Changes
1. Upgrade of recommended Node version from 16 to 18, NPM from 8 to 9
4. Improved README format
5. Enabled `null` entries for baseDB
6. Created functions to create, get, and update entries in execution-table
7. `learn-mod` navigation fixes
8. Begin changing const to functions where possible except for anonymous/callback functions
9. Compacted formal parameters for tabular train functions, to be expanded to other training functions for code conciseness

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,45 +3,30 @@
...
-```
-npm run startb
-```
+## To start on localhost:
+| Action                                                   | Command                |
+| -------------------------------------------------------- | ---------------------- |
+| Install / Update Frontend Packages (one-time)            | `npm run installf`     |
+| Install / Update Backend Packages (one-time) | `npm run installb` |
+| Running the Frontend                                     | `npm run startf`       |
+| Running the Backend                         | `npm run startb-mac`   |
```
Comment (karkir0003) [inline README.md:27]: why did this command change to `npm run startb-mac`?

File: backend/aws_helpers/dynamo_db_utils/execution_db.py [hunk 2]
```diff
@@ -31,4 +36,53 @@ class ExecutionDDBUtil(BaseDDBUtil):
...
+    return ExecutionDDBUtil(EXECUTION_TABLE_NAME, region)
+
+def getOrCreateUserExecutionsData_(entryData: dict) -> str:
+    """
+    Retrieves an entry from the `execution-table` DynamoDB table given an `execution_id`. If does not exist, create a new entry corresponding to the given user_id.
+
+    @param **kwargs: execution_id and other table attributes to be created to the new entry e.g. user_id, if does not exist
+    @return: A JSON string of the entry retrieved or created from the table
+    """
+    dynamoTable = ExecutionDDBUtil(EXECUTION_TABLE_NAME, AWS_REGION)
```
Comment (karkir0003) [inline backend/aws_helpers/dynamo_db_utils/execution_db.py:48]: redundant

Comment (karkir0003) [reply backend/aws_helpers/dynamo_db_utils/execution_db.py:48]: you can replace with `get_execution_table`. We want to sort of enforce singleton here

File: backend/aws_helpers/dynamo_db_utils/execution_db.py [hunk 3]
```diff
@@ -31,4 +36,53 @@ class ExecutionDDBUtil(BaseDDBUtil):
...
+    dataInput = {
+        "data_source": requestData.get("data_source", None),
+        "name": requestData.get("name", None),
+        "progress": requestData.get("progress", None),
+        "status": requestData.get("status", None),
+        "timestamp": requestData.get("timestamp", None),
+        "user_id": requestData["user_id"]
+    }
+
+    dynamoTable = ExecutionDDBUtil(EXECUTION_TABLE_NAME, AWS_REGION)
```
Comment (karkir0003) [inline backend/aws_helpers/dynamo_db_utils/execution_db.py:76]: same idea here

File: backend/common/ai_drive.py [hunk 4]
```diff
@@ -16,26 +17,45 @@
...
-    target=None,
-    features=None,
-    default=None,
-    test_size=0.2,
-    epochs=5,
-    shuffle=True,
-    batch_size=20,
-    json_csv_data_str=""
+    user_arch: list,
+    fileURL: str,
```
Comment (karkir0003) [inline backend/common/ai_drive.py:25]: dont forget to consider default dataset case here

Comment (karkir0003) [reply backend/common/ai_drive.py:25]: do we rlly need json_csv_data_str considering thing with s3? thatll be for another iteration, not for this pr

File: backend/driver.py [hunk 5]
```diff
@@ -45,41 +46,37 @@ def root(path):
     else:
         return send_from_directory(app.static_folder, "index.html")
 
+@app.route("/test")
+def verify_backend_alive():
```
Comment (karkir0003) [inline backend/driver.py:50]: is this used anywhere?

File: frontend/playground-frontend/src/components/helper_functions/TalkWithBackend.js [hunk 6]
```diff
@@ -2,17 +2,20 @@ import { toast } from "react-toastify";
...
-};
+}
 
-const sendToBackend = async (route, data) => {
+async function sendToBackend(route, data) {
   let headers = auth.currentUser
-    ? { Authorization: "bearer " + (await auth.currentUser.getIdToken(true)) }
+    ? {
+        Authorization: "bearer " + (await auth.currentUser.getIdToken(true)),
+        uid: auth.currentUser.uid,
```
Comment (karkir0003) [inline frontend/playground-frontend/src/components/helper_functions/TalkWithBackend.js:17]: what makes this function different from previously?

**Review by karkir0003 — CHANGES_REQUESTED** [review]
address comments. overall, great description of your pr. had some questions @farisdurrani

File: README.md (hunk 1 above)
Comment (farisdurrani) [reply README.md:27]: That's a mistake. Good catch

File: backend/common/ai_drive.py (hunk 4 above)
Comment (farisdurrani) [reply backend/common/ai_drive.py:25]: The default cases have been considered. json_csv_data was already here before. I don't want tk remove it until we get the S3 thing working

File: backend/driver.py (hunk 5 above)
Comment (farisdurrani) [reply backend/driver.py:50]: Not really, this is for debugging purposes to ensure the backend URL is reachable. You'd put the URL into your browser and verify the backend is alive and connected

Comment (farisdurrani) [reply backend/driver.py:50]: I'd say it is needed

File: frontend/playground-frontend/src/components/helper_functions/TalkWithBackend.js (hunk 6 above)
Comment (farisdurrani) [reply frontend/playground-frontend/src/components/helper_functions/TalkWithBackend.js:17]: Previously the header only contains the auth token but now the header contains both the token and UID

**Review by karkir0003 — CHANGES_REQUESTED** [review]
respond to followup comments

File: backend/aws_helpers/dynamo_db_utils/execution_db.py [hunk 7]
```diff
@@ -31,4 +36,53 @@ class ExecutionDDBUtil(BaseDDBUtil):
 
 def get_execution_table(region:str = AWS_REGION) -> BaseDDBUtil:
```
Comment (karkir0003) [inline backend/aws_helpers/dynamo_db_utils/execution_db.py:37]: is `get_execution_table` ever used or referenced anywhere else? if not, then it's not needed

**Review by karkir0003 — CHANGES_REQUESTED** [review]
address remaining comments

File: backend/aws_helpers/dynamo_db_utils/execution_db.py [hunk 8]
```diff
@@ -1,17 +1,22 @@
...
-    timestamp: str
-    data_source: str
-    status: str
-    progress: int
+    user_id: str = None
+    name: str = None
+    timestamp: str = None
+    data_source: str = None
+    status: str = None
+    progress: int = None
```
Comment (farisdurrani) [inline backend/aws_helpers/dynamo_db_utils/execution_db.py:19]: I have to leave all but execution_id as nullable to allow getting an entry using only an execution_id

**Comment by farisdurrani** [comment]
All comments addressed @karkir0003

File: backend/aws_helpers/dynamo_db_utils/execution_db.py (hunk 8 above)
Comment (karkir0003) [reply backend/aws_helpers/dynamo_db_utils/execution_db.py:19]: understood. that's ok

**Review by karkir0003 — APPROVED** [review]

**Pull request merged**